File: 2023-05-05/F/F.go
// https://codeforces.com/problemset/problem/1791/E

// CodeForces 1791E - Negatives and Positives

// Given an array a consisting of n elements, find the maximum possible sum the array can have after performing the following operation any number of times:
//  • Choose 2 adjacent elements and flip both of their signs. In other words choose an index i such that 1≤i≤n−1 and assign ai=−ai and ai+1=−ai+1.

// Input
// The input consists of multiple test cases. The first line contains an integer t (1≤t≤1000) — the number of test cases. The descriptions of the test cases follow.
// The first line of each test case contains an integer n (2≤n≤2⋅10^5) — the length of the array.
// The following line contains n space-separated integers a1,a2,…,an (−10^9≤ai≤10^9).
// It is guaranteed that the sum of n over all test cases does not exceed 2⋅10^5.

// Output
// For each test case, output the maximum possible sum the array can have after performing the described operation any number of times.

// Example
// input
// 5
// 3
// -1 -1 -1
// 5
// 1 5 -5 0 2
// 3
// 1 2 3
// 6
// -1 10 9 8 7 6
// 2
// -1 -1
// output
// 1
// 13
// 6
// 39
// 2

// Note
// For the first test case, by performing the operation on the first two elements, we can change the array from [−1,−1,−1] to [1,1,−1], and it can be proven this array obtains the maximum possible sum which is 1+1+(−1)=1.
// For the second test case, by performing the operation on −5 and 0, we change the array from [1,5,−5,0,2] to [1,5,−(−5),−0,2]=[1,5,5,0,2], which has the maximum sum since all elements are non-negative. So, the answer is 1+5+5+0+2=13.
// For the third test case, the array already contains only positive numbers, so performing operations is unnecessary. The answer is just the sum of the whole array, which is 1+2+3=6.

// Tutorial
// We can notice that by performing any number of operations, the parity of the count of negative numbers won't ever change. Thus, if the number of negative numbers is initially even, we can make it equal to 0 by performing some operations. So, for an even count of negative numbers, the answer is the sum of the absolute values of all numbers (since we can make all of them positive). And if the count of negative numbers is odd, we must have one negative number at the end. We will choose the one smallest by absolute value and keep the rest positive (for simplicity, we consider −0 as a negative number).

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	var reader = bufio.NewReader(os.Stdin)
	var writer = bufio.NewWriter(os.Stdout)
	var c int
	var t, n, i, sum, min uint
	var a []int
	var flag bool

	for c, _ = fmt.Fscan(reader, &t); c == 1; c, _ = fmt.Fscan(reader, &t) {
		for ; t > 0; t-- {
			fmt.Fscan(reader, &n)
			a = make([]int, n)
			sum = 0
			flag = false
			min = 1e9

			for i = 0; i < n; i++ {
				fmt.Fscan(reader, &a[i])

				if a[i] < 0 {
					flag = !flag
					a[i] = -a[i]
				}

				sum += uint(a[i])

				if min > uint(a[i]) {
					min = uint(a[i])
				}
			}

			if flag {
				sum -= min << 1
			}

			fmt.Fprintln(writer, sum)
		}
	}

	writer.Flush()
}

File: 2023-06-02/F/F.cpp
// https://codeforces.com/problemset/problem/1433/G
// You are a mayor of Berlyatov. There are n districts and m two-way roads between them. The i-th road connects districts xi and yi. The cost of travelling along this road is wi. There is some path between each pair of districts, so the city is connected.
// There are k delivery routes in Berlyatov. The i-th route is going from the district ai to the district bi. There is one courier on each route and the courier will always choose the cheapest (minimum by total cost) path from the district ai to the district bi to deliver products.
// The route can go from the district to itself, some couriers routes can coincide (and you have to count them independently).
// You can make at most one road to have cost zero (i.e. you choose at most one road and change its cost with 0).
// Let d(x,y) be the cheapest cost of travel between districts x and y.
// Your task is to find the minimum total courier routes cost you can achieve, if you optimally select the some road and change its cost with 0. In other words, you have to find the minimum possible value of ∑i=1kd(ai,bi) after applying the operation described above optimally.

// Input
// The first line of the input contains three integers n, m and k (2≤n≤1000; n−1≤m≤min(1000,n(n−1)2); 1≤k≤1000) — the number of districts, the number of roads and the number of courier routes.
// The next m lines describe roads. The i-th road is given as three integers xi, yi and wi (1≤xi,yi≤n; xi≠yi; 1≤wi≤1000), where xi and yi are districts the i-th road connects and wi is its cost. It is guaranteed that there is some path between each pair of districts, so the city is connected. It is also guaranteed that there is at most one road between each pair of districts.
// The next k lines describe courier routes. The i-th route is given as two integers ai and bi (1≤ai,bi≤n) — the districts of the i-th route. The route can go from the district to itself, some couriers routes can coincide (and you have to count them independently).

// Output
// Print one integer — the minimum total courier routes cost you can achieve (i.e. the minimum value ∑i=1kd(ai,bi), where d(x,y) is the cheapest cost of travel between districts x and y) if you can make some (at most one) road cost zero.

// Examples

// input
// 6 5 2
// 1 2 5
// 2 3 7
// 2 4 4
// 4 5 2
// 4 6 8
// 1 6
// 5 3
// output
// 22

// input
// 5 5 4
// 1 2 5
// 2 3 4
// 1 4 3
// 4 3 7
// 3 5 2
// 1 5
// 1 3
// 3 3
// 1 5
// output
// 13

// Note
// The picture corresponding to the first example:
// There, you can choose either the road (2,4) or the road (4,6). Both options lead to the total cost 22.
// The picture corresponding to the second example:
// There, you can choose the road (3,4). This leads to the total cost 13.

// Tutorial
// If we would naively solve the problem, we would just try to replace each edge's cost with zero and run Dijkstra algorithm n times to get the cheapest paths. But this is too slow.
// Let's try to replace each edge's cost with zero anyway but use some precalculations to improve the speed of the solution. Let's firstly run Dijkstra n times to calculate all cheapest pairwise paths. Then, let's fix which edge we "remove" (x,y).
// There are three cases for the path (a,b): this edge was not on the cheapest path before removing and is not on the cheapest path after removing. Then the cost of this path is d(a,b). The second case is when this edge was not on the cheapest path before removing but it is on the cheapest path after removing. Then the cost of this path is min(d(a,x)+d(y,b),d(a,y)+d(x,b)). So we are just going from a to x using the cheapest path, then going through the zero edge and then going from y to b using the cheapest path also (or vice versa, from a to y and from x to b). And the third case is when this edge was already on the cheapest path between a and b but this case is essentially the same as the second one.
// So, if we fix the edge (x,y), then the answer for this edge is ∑i=1kmin(d(ai,bi),d(ai,x)+d(y,bi),d(ai,y)+d(x,bi)). Taking the minimum over all edges, we will get the answer.
// The precalculating part works in O(nmlogn) and the second part works in O(km).

#include <iostream>
#include <queue>
#include <set>
#include <utility>
#include <vector>

void dijkstra(const std::vector<std::vector<std::pair<unsigned, unsigned>>> & adj, unsigned s, std::vector<unsigned> & dist) {
    unsigned u, v, w, d, i;
    std::priority_queue<std::pair<unsigned, unsigned>, std::vector<std::pair<unsigned, unsigned>>, std::greater<std::pair<unsigned, unsigned>>> pq;

    dist.resize(adj.size());
    dist.assign(adj.size(), 1e9);
	dist[s] = 0;
	pq.push(std::make_pair(dist[s], s));

	while (!pq.empty()) {
        d = pq.top().first;
		u = pq.top().second;
		pq.pop();

        if (d > dist[u]) {
            continue;
        }

        for (i = 0; i < adj[u].size(); i++) {
            v = adj[u][i].first;
            w = adj[u][i].second;

			if (dist[v] > dist[u] + w) {
                dist[v] = dist[u] + w;
                pq.push(std::make_pair(dist[v], v));
			}
		}
	}
}

int main() {
	unsigned n, m, k, u, v, w, a, b, ans, cur, i, j;
    std::vector<std::vector<std::pair<unsigned, unsigned>>> adj;
    std::vector<std::pair<unsigned, unsigned>> route;
    std::vector<std::vector<unsigned>> d;

	while (std::cin >> n >> m >> k) {
        adj.resize(n);
        adj.assign(n, std::vector<std::pair<unsigned, unsigned>>());
        route.resize(k);
        d.resize(n);
        ans = 1e9;

        for (i = 0; i < m; i++) {
            std::cin >> u >> v >> w;
            u--;
            v--;
            adj[u].push_back({v, w});
            adj[v].push_back({u, w});
        }
        
        for (i = 0; i < k; i++) {
            std::cin >> a >> b;
            a--;
            b--;
            route[i] = std::make_pair(a, b);
        }

        for (u = 0; u < n; u++) {
            dijkstra(adj, u, d[u]);
        }
        
        for (u = 0; u < n; u++) {
            for (i = 0; i < adj[u].size(); i++) {
                v = adj[u][i].first;
                w = adj[u][i].second;
                cur = 0;

                for (j = 0; j < route.size(); j++) {
                    a = route[j].first;
                    b = route[j].second;

                    if (d[a][b] <= d[a][u] + d[v][b] && d[a][b] <= d[a][v] + d[u][b]) {
                        cur += d[a][b];
                    } else if (d[a][u] + d[v][b] <= d[a][v] + d[u][b]) {
                        cur += d[a][u] + d[v][b];
                    } else {
                        cur += d[a][v] + d[u][b];
                    }
                }

                if (ans > cur) {
                    ans = cur;
                }
            }
        }
        
        std::cout << ans << std::endl;
    }
	
	return 0;
}
